cpp/src/arrow/ipc/adapter.cc

```diff
@@ -243,6 +271,26 @@ class RowBatchReader::Impl {
       return MakePrimitiveArray(
           type, field_meta.length, data, field_meta.null_count, null_bitmap, out);
     }
+    if (IsListType(type.get())) {
```

Comment:
The prior block is executed if `IsPrimitive(...)` is true -- in theory they should be mutually exclusive